File: HRForz-API-Node/src/core/storage/s3.ts
import {
  S3Client,
  PutObjectCommand,
  DeleteObjectCommand,
  GetObjectCommand,
} from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import { v4 as uuidv4 } from 'uuid';
import { config } from '../../config';
import { logger } from '../logging/logger';

export const SINGLE_FILE_CATEGORIES = new Set([
  'offer_letter',
  'pan_card',
  'aadhaar_card',
  'bank_passbook',
  'relieving_letter',
]);

class S3StorageClient {
  private client: S3Client;
  private bucket: string;

  constructor() {
    const clientConfig: any = {
      region: config.AWS_S3_REGION,
      credentials: {
        accessKeyId: config.AWS_ACCESS_KEY_ID,
        secretAccessKey: config.AWS_SECRET_ACCESS_KEY,
      },
    };

    if (config.AWS_S3_ENDPOINT_URL) {
      clientConfig.endpoint = config.AWS_S3_ENDPOINT_URL;
      clientConfig.forcePathStyle = true;
    }

    this.client = new S3Client(clientConfig);
    this.bucket = config.AWS_S3_BUCKET;
  }

  async upload(fileBuffer: Buffer, key: string, contentType: string): Promise<string> {
    try {
      await this.client.send(
        new PutObjectCommand({
          Bucket: this.bucket,
          Key: key,
          Body: fileBuffer,
          ContentType: contentType,
        })
      );
      return key;
    } catch (err) {
      logger.error({ error: (err as Error).message, key }, 'storage_upload_failed');
      throw err;
    }
  }

  async getPresignedUrl(key: string, expiresIn: number = 3600): Promise<string> {
    try {
      const command = new GetObjectCommand({
        Bucket: this.bucket,
        Key: key,
      });
      return await getSignedUrl(this.client, command, { expiresIn });
    } catch (err) {
      logger.error({ error: (err as Error).message, key }, 'storage_presign_failed');
      throw err;
    }
  }

  async delete(key: string): Promise<void> {
    try {
      await this.client.send(
        new DeleteObjectCommand({
          Bucket: this.bucket,
          Key: key,
        })
      );
    } catch (err) {
      logger.error({ error: (err as Error).message, key }, 'storage_delete_failed');
      throw err;
    }
  }

  static buildKey(folder: string, filename: string): string {
    const safeName = `${uuidv4()}_${filename}`;
    return `${folder}/${safeName}`;
  }

  static buildDocumentKey(employeeCode: string, category: string, filename: string): string {
    const ts = new Date().toISOString().replace(/[-:T]/g, '').slice(0, 15);
    const safe = filename.replace(/[^\w.\-]/g, '_');
    return `employees/${employeeCode}/documents/${category}/${ts}_${safe}`;
  }
}

export const storage = new S3StorageClient();
export { S3StorageClient };

File: HRForz-API/app/core/storage/s3.py
import re
import uuid
from datetime import datetime, timezone

import boto3
import structlog
from botocore.client import Config
from botocore.exceptions import ClientError

from app.config import settings
from app.core.storage.base import StorageClient

log = structlog.get_logger()

# Categories where only one document is allowed (uploading replaces the existing one)
SINGLE_FILE_CATEGORIES = {
    "offer_letter", "pan_card", "aadhaar_card",
    "bank_passbook", "relieving_letter",
}


class S3StorageClient(StorageClient):
    def __init__(self) -> None:
        kwargs: dict = {
            "aws_access_key_id": settings.AWS_ACCESS_KEY_ID,
            "aws_secret_access_key": settings.AWS_SECRET_ACCESS_KEY,
            "region_name": settings.AWS_S3_REGION,
            "config": Config(signature_version="s3v4"),
        }
        if settings.AWS_S3_ENDPOINT_URL:
            kwargs["endpoint_url"] = settings.AWS_S3_ENDPOINT_URL

        self._client = boto3.client("s3", **kwargs)
        self._bucket = settings.AWS_S3_BUCKET

    async def upload(self, file_bytes: bytes, key: str, content_type: str) -> str:
        try:
            self._client.put_object(
                Bucket=self._bucket,
                Key=key,
                Body=file_bytes,
                ContentType=content_type,
            )
            return key
        except ClientError as exc:
            log.error("storage_upload_failed", key=key, error=str(exc))
            raise

    async def get_presigned_url(self, key: str, expires_in: int = 3600) -> str:
        try:
            return self._client.generate_presigned_url(
                "get_object",
                Params={"Bucket": self._bucket, "Key": key},
                ExpiresIn=expires_in,
            )
        except ClientError as exc:
            log.error("storage_presign_failed", key=key, error=str(exc))
            raise

    async def delete(self, key: str) -> None:
        try:
            self._client.delete_object(Bucket=self._bucket, Key=key)
        except ClientError as exc:
            log.error("storage_delete_failed", key=key, error=str(exc))
            raise

    @staticmethod
    def build_key(folder: str, filename: str) -> str:
        safe_name = f"{uuid.uuid4()}_{filename}"
        return f"{folder}/{safe_name}"

    @staticmethod
    def build_document_key(employee_code: str, category: str, filename: str) -> str:
        """Organized per-employee folder: employees/{code}/documents/{category}/{ts}_{name}"""
        ts = datetime.now(timezone.utc).strftime("%Y%m%d_%H%M%S")
        safe = re.sub(r"[^\w.\-]", "_", filename)
        return f"employees/{employee_code}/documents/{category}/{ts}_{safe}"


storage: S3StorageClient = S3StorageClient()
